### PR TITLE
Encode labels in the tf.data pipeline for Object Detection models

### PR DESCRIPTION
**Pull request by LukeWood**
Draft PR, please do not review yet.

### CODE REVIEW

File: keras_cv/models/object_detection/object_detection_base_model.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+# limitations under the License.
+
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_cv import bounding_box
+
+
+class ObjectDetectionBaseModel(keras.Model):
+
```
Comment (fchollet) [inline keras_cv/models/object_detection/object_detection_base_model.py:22]: Remove blank line

File: keras_cv/models/object_detection/object_detection_base_model.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        self.bounding_box_format = bounding_box_format
+        self.label_encoder = label_encoder
+        self.built = True
+
+    def fit(self, x=None, y=None, sample_weight=None, batch_size=None, **kwargs):
+        if sample_weight is not None:
+            raise ValueError("RetinaNet does not yet support `sample_weight`.")
+
+        dataset = _convert_inputs_to_tf_dataset(x=x, y=y, batch_size=batch_size)
+        dataset = dataset.map(self.encode_data, num_parallel_calls=tf.data.AUTOTUNE)
```
Comment (fchollet) [inline keras_cv/models/object_detection/object_detection_base_model.py:43]: Please check with the tf.data team about the need to potentially re-prefetch after `map`

File: keras_cv/models/object_detection/object_detection_base_model.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    def __init__(self, bounding_box_format, label_encoder, **kwargs):
+        super().__init__(**kwargs)
+        self.bounding_box_format = bounding_box_format
+        self.label_encoder = label_encoder
+
+        self.label_encoder.build((None, None, None))
+
+    def fit(self, x=None, y=None, sample_weight=None, batch_size=None, **kwargs):
+        if sample_weight is not None:
+            raise ValueError("RetinaNet does not yet support `sample_weight`.")
```
Comment (fchollet) [inline keras_cv/models/object_detection/object_detection_base_model.py:40]: We may want to add it -- it should be relatively straightforward. Not a must before shipping though.

File: keras_cv/models/object_detection/object_detection_base_model.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    """
+
+    def __init__(self, bounding_box_format, label_encoder, **kwargs):
+        super().__init__(**kwargs)
+        self.bounding_box_format = bounding_box_format
+        self.label_encoder = label_encoder
+
+        self.label_encoder.build((None, None, None))
+
+    def fit(self, x=None, y=None, sample_weight=None, batch_size=None, **kwargs):
```
Comment (fchollet) [inline keras_cv/models/object_detection/object_detection_base_model.py:38]: Having support for validation_data and validation_split is a must before shipping

File: keras_cv/models/object_detection/object_detection_base_model.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/object_detection_base_model.py]: lets do it as a follow up - I have to support it in the label encoder too.

**Comment by LukeWood** [comment]
@fchollet I believe this PR is ready!

File: keras_cv/models/object_detection/object_detection_base_model.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/object_detection_base_model.py:68]: Have not yet had contact with tf.data, but I did verify that my pipeline gets 98% GPU utilization.

Comment (LukeWood) [reply keras_cv/models/object_detection/object_detection_base_model.py:68]: any reason to not just always prefetch anyways though?  Perhaps this will give us a small speedup regardless.

File: keras_cv/models/object_detection/__internal__.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+import functools
+import math
+
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+
+
+def _split_validation_data(validation_data):
```
Comment (fchollet) [inline keras_cv/models/object_detection/__internal__.py:23]: This function can be replaced by `keras.utils.unpack_x_y_sample_weight`.

File: keras_cv/models/object_detection/__internal__.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import math
+
+import numpy as np
+import pandas as pd
```
Comment (fchollet) [inline keras_cv/models/object_detection/__internal__.py:19]: Rather than requiring `pandas` (which is heavy) just to support the optional use case of passing a dataframe, I recommend making it an optional dependency, like https://github.com/keras-team/keras/blob/master/keras/engine/data_adapter.py#L42-L45

File: keras_cv/models/object_detection/object_detection_base_model.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+from keras_cv.models.object_detection.__internal__ import _split_validation_data
+from keras_cv.models.object_detection.__internal__ import _train_validation_split
+
+
+class ObjectDetectionBaseModel(keras.Model):
+    """ObjectDetectionBaseModel performs asynchonous label encoding.
+
+    ObjectDetectionBaseModel invokes the provided `label_encoder` in the `tf.data`
+    pipeline to ensure optimal training performance.  This is done by overriding the
+    methods `train_on_batch()`, `fit()`, `test_on_batch()`, and `evaluate()`.
```
Comment (fchollet) [inline keras_cv/models/object_detection/object_detection_base_model.py:29]: This PR does not include the `on_batch` methods, please add them.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

**Comment by LukeWood** [comment]
ok @fchollet addressed all comments!

**Review by fchollet — APPROVED** [review]
LGTM, thanks!

**Comment by LukeWood** [comment]
FYI @martin-gorner this has been merged!  Feel free to test on TPU :)
